internal/librarian/update_image.go

```diff
@@ -0,0 +1,190 @@
+	librarianConfig *config.LibrarianConfig
+	repo            gitrepo.Repository
+	sourceRepo      gitrepo.Repository
+	state           *config.LibrarianState
+	build           bool
+	push            bool
+	commit          bool
+	image           string
+	workRoot        string
+}
+
+// ImageRegistryClient is an abstraction around interacting with image.
+type ImageRegistryClient interface {
+	FindLatest(ctx context.Context, imageName string) (string, error)
+}
+
+func newUpdateImageRunner(cfg *config.Config) (*updateImageRunner, error) {
+	runner, err := newCommandRunner(cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &updateImageRunner{
+		branch:          cfg.Branch,
+		containerClient: runner.containerClient,
+		ghClient:        runner.ghClient,
+		hostMount:       cfg.HostMount,
+		librarianConfig: runner.librarianConfig,
+		repo:            runner.repo,
+		sourceRepo:      runner.sourceRepo,
+		state:           runner.state,
+		build:           cfg.Build,
+		commit:          cfg.Commit,
+		push:            cfg.Push,
+		image:           cfg.Image,
+		workRoot:        runner.workRoot,
+	}, nil
+}
+
+func (r *updateImageRunner) run(ctx context.Context) error {
+	imagesClient := r.imagesClient
+	if imagesClient == nil {
+		slog.Info("no imagesClient provided, defaulting to ArtifactRegistry implementation")
+		client, err := images.NewArtifactRegistryClient(ctx)
+		if err != nil {
+			return err
+		}
+		defer client.Close()
+		imagesClient = client
+	}
+
+	// Update `image` entry in state.yaml
+	if r.image == "" {
+		slog.Info("No image found, looking up latest")
+		latestImage, err := imagesClient.FindLatest(ctx, r.state.Image)
+		if err != nil {
+			slog.Error("Unable to determine latest image to use", "image", r.state.Image)
+			return err
+		}
+		r.image = latestImage
+	}
+
+	if r.image == r.state.Image {
+		slog.Info("No update to the image, aborting.")
+		return nil
+	}
+
+	r.state.Image = r.image
+
+	if err := saveLibrarianState(r.repo.GetDir(), r.state); err != nil {
+		return err
+	}
+
+	// For each library, run generation at the previous commit
+	failedGenerations := make([]*config.LibraryState, 0)
+	successfulGenerations := make([]*config.LibraryState, 0)
+	outputDir := filepath.Join(r.workRoot, "output")
+	for _, libraryState := range r.state.Libraries {
+		err := r.regenerateSingleLibrary(ctx, libraryState, outputDir)
+		if err != nil {
+			slog.Error(err.Error(), "library", libraryState.ID, "commit", libraryState.LastGeneratedCommit)
+			failedGenerations = append(failedGenerations, libraryState)
+			continue
+		} else {
+			successfulGenerations = append(successfulGenerations, libraryState)
+		}
+	}
+	slog.Warn("failed generations", slog.Int("num", len(failedGenerations)))
+	slog.Info("successful generations", slog.Int("num", len(successfulGenerations)))
+
+	// TODO(#2587): improve PR body content
+	prBodyBuilder := func() (string, error) {
+		body := fmt.Sprintf("feat: update image to %s\n\n", r.image)
```

Comment:
For all of our other PR bodies we use go templates. I think it would be nice to do the same here for readability and consistency.

Reply:
SG